File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/SemanticLexicalCandidate.java
package com.foundgine.core.semantic.resolution;

import com.foundgine.core.abstractions.*;
import java.util.List;

public record SemanticLexicalCandidate(
    String token, SemanticLexicalCandidateKind kind, String canonicalName, double score,
    EntityId entityId, RelationshipId relationshipId, FieldId fieldId,
    EntityId sourceEntityId, EntityId targetEntityId, String value,
    List<ResolutionEvidence> evidence) {
    public SemanticLexicalCandidate {
        evidence = evidence == null ? List.of() : List.copyOf(evidence);
    }
    public SemanticLexicalCandidate(String token, SemanticLexicalCandidateKind kind, String canonicalName, double score) {
        this(token, kind, canonicalName, score, null, null, null, null, null, null, List.of());
    }
    public List<ResolutionEvidence> effectiveEvidence() { return evidence; }
}

File: src/csharp/Foundgine.Core/Semantic/Resolution/Lexical/SemanticLexicalCandidate.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Resolution;

/// <summary>
/// A lexical hypothesis returned by an approximate retrieval provider.
/// Score is provider relevance, not a probability and not an authorization
/// decision. Foundgine combines it with semantic graph compatibility.
/// </summary>
public sealed record SemanticLexicalCandidate(
    string Token,
    SemanticLexicalCandidateKind Kind,
    string CanonicalName,
    double Score,
    EntityId? EntityId = null,
    RelationshipId? RelationshipId = null,
    FieldId? FieldId = null,
    EntityId? SourceEntityId = null,
    EntityId? TargetEntityId = null,
    string? Value = null,
    IReadOnlyList<ResolutionEvidence>? Evidence = null)
{
    public IReadOnlyList<ResolutionEvidence> EffectiveEvidence => Evidence ?? [];
}
